src/ai/defaultai.cc

```diff
@@ -1297,125 +1297,130 @@ void DefaultAI::late_initialization() {
-				if (maxsize > 1) {
-					--non_small_needed;
-				}
-			}
-			if (near_border_needed > 0) {
-				if (bf.near_border) {
-					--near_border_needed;
-				}
-			}
+		} else {  // just rotating
+			buildable_fields.push_back(&bf);
+			buildable_fields.pop_front();
 		}
-		bf.field_info_expiration = gametime + kFieldInfoExpiration;
-		buildable_fields.push_back(&bf);
-		buildable_fields.pop_front();
-
-		++i;
-	}
-
-	// If needed we iterate once more and look for 'special' fields
-	// starting in the middle of buildable_fields to skip fields tested lately
-	// But not doing this if the count of buildable fields is too low
-	// (no need to bother)
-	if (buildable_fields.size() < kMinimalFieldsCheck * 3) {
-		return;
 	}
 
-	for (uint32_t j = buildable_fields.size() / 2; j < buildable_fields.size(); j++) {
-		// If we dont need to iterate (anymore) ...
-		if (non_small_needed + near_border_needed == 0) {
+	// Stage #3: update all buildable fields (expired ones of course) up to the limit
+	for (uint32_t j = 0; j < buildable_fields.size(); j++) {
+		if (updated_fields_count >= max_fields_to_check) {
 			break;
 		}
-
-		// Skip if the field is not ours or was updated lately
-		if (buildable_fields[j]->coords.field->get_owned_by() != player_number()) {
-			continue;
-		}
-		// We are not interested in fields where info has expired less than 20s ago
-		if (buildable_fields[j]->field_info_expiration + Duration(20 * 1000) > gametime) {
-			continue;
-		}
-
-		// Continue if field is blocked at the moment
-		if (blocked_fields.is_blocked(buildable_fields[j]->coords)) {
-			continue;
-		}
-
-		// Few constants to keep the code cleaner
-		const int32_t field_maxsize =
-		   player_->get_buildcaps(buildable_fields[j]->coords) & Widelands::BUILDCAPS_SIZEMASK;
-		const bool field_near_border = buildable_fields[j]->near_border;
-
-		// Let decide if we need to update and for what reason
-		const bool update_due_size = non_small_needed && field_maxsize > 1;
-		const bool update_due_border = near_border_needed && field_near_border;
-
-		if (!(update_due_size || update_due_border)) {
-			continue;
+		assert(!buildable_fields[j]->invalidated);  // there should be no more invalid fields
+		if (buildable_fields[j]->field_info_expiration < gametime) {
+			update_buildable_field(*buildable_fields[j]);
+			buildable_fields[j]->field_info_expiration = gametime + kFieldInfoExpiration;
+			updated_fields_count++;
 		}
+	}
 
-		// decreasing the counters
-		if (update_due_size) {
-			assert(non_small_needed > 0);
-			--non_small_needed;
-		}
-		if (update_due_border) {
-			assert(near_border_needed > 0);
-			--near_border_needed;
-		}
+	assert(updated_fields_count <= max_fields_to_check);
 
-		// and finnaly update the buildable field
-		update_buildable_field(*buildable_fields[j]);
-		buildable_fields[j]->field_info_expiration = gametime + kFieldInfoExpiration;
-	}
+	verb_log_dbg_time(gametime, " ... %2d fields updated of %4lu.\n", updated_fields_count,
```

Comment:
replace "%4lu" with "%4llu" as the variable is long long unsigned

Reply:
done

Reply:
Actually it doesn't matter with the new `format()` function. The `l`-s are ignored.

Reply:
the mingw compiler is requireing the double "ll" as it seems the do_log fuction is not using the new format function and does require this to be properly specified.

Reply:
so perhaps I will put there:
static_cast<uint32_t>buildable_fields.count()

Reply:
Sorry, that's right. `do_log()` uses `vsnprintf()` not `format()`.

Reply:
Or I can just remove part of log with the number

Reply:
I would go to cast it, if %llu does conmflict in other compilers. But as far as I understood it might be ignored there. so either way would be fine I think.

Reply:
```suggestion
	verb_log_dbg_time(gametime, " ... %2d fields updated of %" PRIuS ".\n", updated_fields_count,
```

size_t is a different type under each compiler, and the PRIuS macro exists to handle them all